Facilitate better backtraces in CI

hmenke opened the pull request.

Wentzell committed:
Squash app4triqs/3.1.x to reduce skeleton history when tracking

Co-authored-by: Dylan Simon <dylan@dylex.net>
Co-authored-by: Alexander Hampel <ahampel@flatironinstitute.org>

Wentzell committed:
[doc] Minor formatting correction in doc/documentation.rst

the-hampel committed:
automatically use version number set by cmake in doc

Wentzell committed:
Set CPLUS_INCLUDE_PATH and LIBRARY_PATH in app4triqsvars.sh

dylex committed:
[jenkins] remove my email

Wentzell committed:
[ghactions] Switch osx build to macos-12

the-hampel committed:
add triqs logos to skeleton and include ico in install directive of doc

Wentzell committed:
[clang-tidy] Do not warn about magic numbers

the-hampel committed:
[build] update github workflow to match triqs unstable req

Wentzell committed:
[ghactions] Update osx build to use gcc 12

Wentzell committed:
[jenkins] Move to gcc 12 for osx builds

Wentzell committed:
Add Wfloat-conversion to warning list

Wentzell committed:
Synchronize clang-tidy config file with triqs

Wentzell committed:
[jenkins] Move to gcc 13 for osx builds

Wentzell committed:
[doc] Update sphinxext.numpydoc.plot_directive to latest version

Wentzell committed:
[doc] Update conf.py to not show source-links and format links by default

Wentzell committed:
[doc] Include plot source by default

Wentzell committed:
[jenkins] Remove redundant '-DBuild_Deps=Always' from cmake configuration

Wentzell committed:
[cmake] Bump Version to 3.2.0, require triqs v3.2, raise cmake req to 3.20

Wentzell committed:
[gitignore] Ignore doc/_autosummary directory

Wentzell committed:
Rename easybuild file and update conda meta yaml with version number

Wentzell committed:
[ghactions] Fix llvm version number for ubuntu env

Wentzell committed:
Disable C++20 compat warnings

Wentzell committed:
[cmake] Include build options for TSAN and MSAN

Wentzell committed:
[cmake] Extend extract_flags.cmake to include cxx_std_23

Wentzell committed:
Do not build documentation for subprojects

Wentzell committed:
[cmake] Do not check doc build options as subproject

the-hampel committed:
Update README.md

Wentzell committed:
[cmake] Install bash files contained in doc pages

Wentzell committed:
[cmake] Protect against redefining santizier targets

the-hampel committed:
[build] add libcxx flags for clang workflows

Wentzell committed:
[cmake] Do not find find cpp2py when PythonSupport==OFF and subproject

Thoemi09 committed:
[cmake] Add compiler warnings for IntelLLVM

Wentzell committed:
[cmake] Fix sanitizer target names

Thoemi09 committed:
[cmake] Change target for compiler warnings

Wentzell committed:
Use python3 instead of python2 in replace_and_rename.py script

Wentzell committed:
Skip image and binary files in replace_and_rename

Wentzell committed:
Raise shm size for docker run commands to comply with mkl requirements

Wentzell committed:
Remove numpydoc sources from doc/sphinxext

the-hampel committed:
[build] automatically set version in packaging

Wentzell committed:
[jenkins] For sanitized build use RelWithDebInfo build mode

Wentzell committed:
[cmake] Run Debug checks also in RelWithDebInfo build mode

the-hampel committed:
[build] add packaging directory to cmake

hmenke commented on .github/workflows/build.yml:
```diff
@@ -96,6 +97,8 @@ jobs:
     - name: Test app4triqs
       env:
         DYLD_FALLBACK_LIBRARY_PATH: /usr/local/opt/llvm/lib
+        LD_PRELOAD: /usr/lib/x86_64-linux-gnu/libSegFault.so
```
Using `libSegFault` is debatable.

Wentzell replied:
Would this have any implications in the absence of segfaults?

hmenke replied:
Essentially the entire constructor boils down to `sigaction (SIGSEGV, &sa, NULL);` where `sa` is the custom segfault handler. So in the absence of `SIGSEGV` this will still consume one additional slot on the signal stack and might clash or otherwise interact with other signal handlers, although I don't think that TRIQS uses POSIX signals for anything.

hmenke committed:
Use ccache to speed up compilation

hmenke committed:
Prevent unintentional parallelization in OpenBLAS

hmenke committed:
Allow manual dispatch and triggering action from other workflow

hmenke committed:
Fix settings environment variables

hmenke committed:
Bump actions/checkout from 2 to 4

Bumps [actions/checkout](https://github.com/actions/checkout) from 2 to 4.
- [Release notes](https://github.com/actions/checkout/releases)
- [Changelog](https://github.com/actions/checkout/blob/main/CHANGELOG.md)
- [Commits](https://github.com/actions/checkout/compare/v2...v4)

---
updated-dependencies:
- dependency-name: actions/checkout
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: Henri Menke <henri@henrimenke.de>

hmenke committed:
Dump the Python traceback on SIGSEGV

https://docs.python.org/3/library/faulthandler.html

hmenke committed:
Compile with debugging symbols

This still includes optimizations. Roughly corresponds to "-O2 -g -DNDEBUG"

https://cmake.org/cmake/help/latest/variable/CMAKE_BUILD_TYPE.html
https://gitlab.kitware.com/cmake/cmake/-/blob/master/Modules/Compiler/GNU.cmake

hmenke committed:
Install and preload libSegFault

Unfortunately there is no upstream documentation, so here are two blogposts.

https://www.marcusfolkesson.se/blog/libsegfault/
https://lemire.me/blog/2023/05/01/under-linux-libsegfault-and-addr2line-are-underrated/

hmenke commented:
I've split the PR into multiple commits, so you can cherry-pick what you consider reasonable.
